src/analysis/normed_space/operator_norm.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2019 Jan-David Salchow. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jan-David Salchow
+
+The space of bounded linear maps
+
+Define the set of bounded linear maps between normed spaces and show basic facts about it. In
+particular
+
+(*) define a set L(E,F) of bounded linear maps between k normed spaces,
+(*) show that L(E,F) is a vector subspace of E → F,
+(*) define the 'operator norm' on L(E,F) and show that it induces the structure of a normed space
+    on L(E,F).
+-/
+import algebra.module
+import analysis.normed_space.bounded_linear_maps
+
+variables {k : Type*} 
+variables {E : Type*} {F : Type*}
+
+
+-- Define the subspace of bounded linear maps, introduce the notation L(E,F) for the set of bounded linear maps.
+section bounded_linear_maps
+
+variables [hnfk : normed_field k] [normed_space k E] [normed_space k F]
+include hnfk
+
+def bounded_linear_maps : subspace k (E → F) :=
+{ carrier := {A : E → F | is_bounded_linear_map A},
+  zero := is_bounded_linear_map.zero,
+  add := assume A B, is_bounded_linear_map.add,
+  smul := assume c A, is_bounded_linear_map.smul c }
+
+notation `L(` E `,` F `)` := @bounded_linear_maps _ E F _ _ _
+
+/-- Coerce bounded linear maps to functions. -/
+instance bounded_linear_maps.to_fun : has_coe_to_fun $ L(E,F) :=
+{F :=  λ _, (E → F), coe := (λ f, f.val)}
+
+@[extensionality] theorem ext {A B : L(E,F)} (H : ∀ x, A x = B x) : A = B :=
+begin
+  cases A with A hA,
+  cases B with B hB,
+  congr',
+  exact funext H
+end
+
+/-- Bounded linear maps are ... bounded -/
+lemma exists_bound (A : L(E,F)) : ∃ c, c > 0 ∧ ∀ x : E, ∥A x∥ ≤ c * ∥x∥ := A.property.bound
+
+/-- Bounded linear maps are conveniently bounded on the unit ball. -/
+lemma exists_bound' (A : L(E,F)) : ∃ c, c > 0 ∧ ∀ x : E, ∥x∥ ≤ 1 → ∥A x∥ ≤ c :=
+let ⟨c, _, H⟩ := exists_bound A in
+exists.intro c ⟨‹c > 0›,
+  assume x _,
+  calc ∥A x∥ ≤ c * ∥x∥ : H x
+        ... ≤ c * 1 : (mul_le_mul_left ‹c > 0›).mpr ‹∥x∥ ≤ 1›
+        ... = c : mul_one c⟩
+
+end bounded_linear_maps
+
+/-
+Now define the operator norm. We only do this for normed spaces over ℝ, since we need a
+scalar multiplication with reals to prove that ∥A x∥ ≤ ∥A∥ * ∥x∥. It would be enough to
+have a vector space over a normed field k with a real scalar multiplication and certain
+compatibility conditions.
+
+The main task is to show that the operator norm is definite, homogeneous, and satisfies the
+triangle inequality. This is done after a few preliminary lemmas necessary to deal with cSup.
+-/
+section operator_norm
+
+variables [normed_space ℝ E] [normed_space ℝ F]
+open lattice set
+
+/-- The operator norm of a bounded linear map A : E → F is the sup of
+    the set ∥A x∥ with ∥x∥ = 1. If E = {0} we set ∥A∥ = 0. -/
+noncomputable def operator_norm : L(E, F) → ℝ :=
+assume A, Sup (image (λ x, ∥A x∥) {x | ∥x∥ ≤ 1})
```

Comment:
```suggestion
λA, ⨆ (x : α) (h : ∥x∥ ≤ 1), ∥A x∥
```

Especially replace the `assume` by `λ`. The `assume` should be only used for proofs, otherwise it gets very confusing.

Reply:
Even better
```lean
noncomputable def operator_norm (A : L(E, F)) : ℝ :=
⨆ (x : α) (h : ∥x∥ ≤ 1), ∥A x∥
```

Reply:
I've changed it to

```lean
noncomputable def operator_norm (A : L(E, F)) : ℝ :=
Sup (image (λ x, ∥A x∥) {x | ∥x∥ ≤ 1})
```

Hopefully, you can live with that?

Reply:
What's wrong with the indexed sup version? It seems clearer and there are theorems for dealing with it without having to unfold image stuff. I would suggest proving equivalence to `Sup ((λ x, ∥A x∥) '' {x | ∥x∥ ≤ 1})` if you need it for some reason

Reply:
My only concern is readability, somehow ⨆ doesn't look like a sup, but like a disjoint union. Can we have notation along the lines of `Sup_[x : α | h : ∥x∥ ≤ 1)]  ∥A x∥` ? Then I'd be happy to change that definition ...